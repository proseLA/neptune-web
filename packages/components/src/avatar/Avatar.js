import classnames from 'classnames';
import PropTypes from 'prop-types';

import { Size } from '../common';

import { AvatarType } from './avatarTypes';

const Avatar = ({ backgroundColor, children, className, outlined, size, type }) => (
  <div
    className={classnames('tw-avatar', className, `tw-avatar--${size}`, `tw-avatar--${type}`, {
      'tw-avatar--outlined': outlined,
    })}
  >
    <div className="tw-avatar__content" style={{ backgroundColor: backgroundColor || undefined }}>
      {children}
    </div>
  </div>
);

Avatar.propTypes = {
  backgroundColor: PropTypes.string,
  children: PropTypes.node,
  className: PropTypes.string,
  outlined: PropTypes.bool,
  size: PropTypes.oneOf(['sm', 'md', 'lg']),
  type: PropTypes.oneOf(['thumbnail', 'icon', 'emoji', 'initials']),
};

Avatar.defaultProps = {
  backgroundColor: null,
  children: null,
  className: null,
  outlined: false,
  size: Size.MEDIUM,
  type: AvatarType.THUMBNAIL,
};

export default Avatar;
