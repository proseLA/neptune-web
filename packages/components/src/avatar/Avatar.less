@import (reference) '@transferwise/neptune-css/src/variables/neptune-tokens.less';
@import (reference) '@transferwise/neptune-css/src/less/mixins/_logical-properties.less';

@avatar-diameter-lg: 72px;
@avatar-diameter-md: 48px;
@avatar-diameter-sm: 24px;

.tw-avatar {
  position: relative;
  border-radius: 50%;
  user-select: none;
  box-sizing: border-box;
  cursor: default;

  & &__content {
    display: flex;
    justify-content: center;
    align-items: center;
    border-radius: 50%;
    overflow: hidden;
    color: var(--color-content-primary);
    background-color: var(--color-background-neutral);
    height: 100%;
    width: 100%;
    max-height: 100%;
    max-width: 100%;
  }

  &--outlined {
    border: 1px solid var(--color-interactive-accent);
    cursor: pointer;

    &:hover {
      border-color: var(--color-interactive-accent-hover);
    }
  }

  // theming: this becomes redundant when we implement theming as we will just change value of CSS custom props
  &--dark &__content {
    background: var(--color-background-neutral);
  }

  &--thumbnail {
    & * {
      width: 100%;
      height: 100%;
      object-fit: cover;
    }
  }

  &--icon {
    & * {
      max-height: 100%;
      max-width: 100%;
    }
  }

  &.tw-avatar--initials,
  &.tw-avatar--icon {
    color: var(--color-content-primary);
  }

  &.tw-avatar--initials,
  &.tw-avatar--emoji {
    // There is a problem how emojis & text vertically aligned between Windows and Mac. Setting font family to Arial seems to resolve it.
    font-family: Arial, sans-serif;
  }

  &--sm {
    .avatar-dimension(@avatar-diameter-sm, 1px);
  }

  &--md {
    .avatar-dimension(@avatar-diameter-md, 2px);
  }

  &--lg {
    .avatar-dimension(@avatar-diameter-lg, 3px);
  }

  img {
    width: 100%;
    height: 100%;
  }
}

.avatar-dimension(@avatar-size, @mask-size) {
  min-width: @avatar-size;
  width: @avatar-size;
  height: @avatar-size;

  &.tw-avatar--emoji,
  &.tw-avatar--icon {
    font-size: @avatar-size * 0.5;
  }

  &.tw-avatar--initials {
    font-size: @avatar-size * 0.35;
    font-weight: var(--font-weight-bold);
  }

  &.tw-avatar--outlined {
    padding: @mask-size;

    &.tw-avatar--thumbnail::after {
      top: @mask-size;
      .left(@mask-size);

      .right(@mask-size);

      bottom: @mask-size;
    }
  }

  &.tw-avatar--thumbnail::after {
    position: absolute;
    top: 0;
    .left(0);

    .right(0);

    bottom: 0;
    border: 1px solid var(--color-border-overlay);
    content: '';
    border-radius: 50%;
  }

  // theming: this becomes redundant when we implement theming as we will just change value of CSS custom props
  &.tw-avatar--dark.tw-avatar--thumbnail::after {
    border-color: rgba(255, 255, 255, 0.08);
  }
}
