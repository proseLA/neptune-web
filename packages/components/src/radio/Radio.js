import classNames from 'classnames';
import PropTypes from 'prop-types';

import RadioButton from '../common/RadioButton';

const Radio = ({ label, id, disabled, className, avatar, secondary, ...otherProps }) => {
  return (
    <div
      className={classNames(
        'radio np-radio',
        {
          'radio-lg': secondary,
        },
        className,
      )}
    >
      <label className={classNames({ disabled })} htmlFor={id}>
        <span className="np-radio-button p-r-2">
          <RadioButton id={id} disabled={disabled} {...otherProps} />
        </span>
        <span className="np-radio__text">
          {label}
          {secondary && <small>{secondary}</small>}
        </span>
        {avatar && <span className="np-radio__avatar m-l-auto">{avatar}</span>}
      </label>
    </div>
  );
};

Radio.propTypes = {
  avatar: PropTypes.element,
  checked: PropTypes.bool,
  disabled: PropTypes.bool,
  id: PropTypes.string,
  label: PropTypes.string.isRequired,
  name: PropTypes.string.isRequired,
  onChange: PropTypes.func.isRequired,
  secondary: PropTypes.string,
  value: PropTypes.oneOfType([PropTypes.number, PropTypes.string]),
  className: PropTypes.string,
};

Radio.defaultProps = {
  avatar: undefined,
  checked: false,
  disabled: false,
  id: null,
  secondary: null,
  value: '',
  className: undefined,
};

export default Radio;
