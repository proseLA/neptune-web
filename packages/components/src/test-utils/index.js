import { render } from '@testing-library/react';
import { renderHook } from '@testing-library/react-hooks';
import userEvent from '@testing-library/user-event';

import { Provider } from '..';
import { DEFAULT_LOCALE } from '../common/locale';
import en from '../i18n/en.json';

/**
 * Custom `render` function which wraps passed elements in Provider component
 * Source: https://testing-library.com/docs/example-react-intl/#creating-a-custom-render-function
 *
 */
function customRender(
  ui,
  { locale = DEFAULT_LOCALE, messages = en, direction, ...renderOptions } = {},
) {
  // eslint-disable-next-line react/prop-types
  var Wrapper = ({ children }) => {
    return (
      <Provider i18n={{ locale, messages }} direction={direction}>
        {children}
      </Provider>
    );
  };
  return render(ui, { wrapper: Wrapper, ...renderOptions });
}

/**
 * Custom `renderHook` function which wraps passed elements in Provider component
 * For more info: https://react-hooks-testing-library.com/usage/advanced-hooks#context
 *
 */
function customRenderHook(callback, { locale = DEFAULT_LOCALE, messages = en, direction } = {}) {
  return renderHook(callback, {
    // eslint-disable-next-line react/prop-types
    wrapper: ({ children }) => (
      <Provider i18n={{ locale, messages }} direction={direction}>
        {children}
      </Provider>
    ),
  });
}

export * from '@testing-library/react';
export { userEvent };
export { customRender as render, customRenderHook as renderHook };
