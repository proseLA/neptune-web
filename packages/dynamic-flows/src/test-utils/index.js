import { Provider, translations as componentTranslations } from '@transferwise/components';
import { mount } from 'enzyme';
import { act } from 'react-dom/test-utils';

import { FetcherProviderFromBaseUrl } from '../common/contexts/fetcherContext';

const getMockFetchPromise = (status, jsonFn, delay, signal) => {
  const response = {
    status,
    json: jsonFn,
  };
  return new Promise((resolve) => {
    let aborted = false;
    if (signal) {
      signal.addEventListener('abort', () => {
        aborted = true;
      });
    }
    setTimeout(() => {
      return !aborted && resolve(response);
    }, delay);
  });
};

const wait = (t) => {
  return act(() => {
    return new Promise((resolve) => setTimeout(resolve, t));
  });
};

function mountWithProviders(
  component,
  locale = 'en-GB',
  // eslint-disable-next-line unicorn/no-object-as-default-parameter
  messages = { ...componentTranslations['en'] },
  baseUrl = 'https://test-url',
) {
  return mount(
    <Provider i18n={{ locale, messages }}>
      <FetcherProviderFromBaseUrl baseUrl={baseUrl}>{component}</FetcherProviderFromBaseUrl>
    </Provider>,
  );
}

function getI18n(
  locale = 'en-GB',
  // eslint-disable-next-line unicorn/no-object-as-default-parameter
  messages = { ...componentTranslations['en'] },
) {
  return { locale, messages };
}

export { mountWithProviders as mount, getMockFetchPromise, wait, getI18n };
