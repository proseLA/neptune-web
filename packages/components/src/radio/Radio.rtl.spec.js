import Avatar, { AvatarType } from '../avatar';
import { render } from '../test-utils';

import Radio from '.';

jest.mock('../common/hooks');

describe('Radio', () => {
  const props = {
    label: 'This is a label',
    name: 'name',
    onChange: jest.fn(),
    checked: false,
    disabled: false,
    secondary: '',
  };

  it('shows the avatar when supplied', () => {
    expect(
      render(<Radio {...props} avatar={<Avatar type={AvatarType.INITIALS}>HD</Avatar>} />)
        .container,
    ).toMatchSnapshot();
  });
});
