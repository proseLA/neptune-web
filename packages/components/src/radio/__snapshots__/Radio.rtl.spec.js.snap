// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Radio shows the avatar when supplied 1`] = `
<div>
  <div
    class="radio np-radio"
  >
    <label
      class=""
    >
      <span
        class="np-radio-button p-r-2"
      >
        <input
          class="sr-only"
          name="name"
          type="radio"
          value=""
        />
        <span
          class="tw-radio-button"
        >
          <span
            class="tw-radio-check"
          />
        </span>
      </span>
      <span
        class="np-radio__text"
      >
        This is a label
        
      </span>
      <span
        class="np-radio__avatar m-l-auto"
      >
        <div
          class="tw-avatar tw-avatar--md tw-avatar--initials tw-avatar--light"
        >
          <div
            class="tw-avatar__content"
          >
            HD
          </div>
        </div>
      </span>
    </label>
  </div>
</div>
`;
